src/egraph.rs

```diff
@@ -202,6 +216,50 @@ impl<L: Language> EGraph<L> {
         pruned
     }
 
+    pub fn fold_constants(&mut self) -> usize {
+        let mut folded = 0;
+        let mut to_add = HashMap::default();
+        let mut constant_nodes = HashMap::default();
```

Comment:
Filling up `constant_nodes` in the loop below might be too late. It introduces an ordering problem. I think you want two passes over the eclasses here. One to add the constants, and another to do the folding.

Reply:
How about the code now?